riscv_core/src/trace.rs

```diff
@@ -0,0 +1,537 @@
+/// RISC-V Instruction Trace Structures
+///
+/// Provides structured representation of executed instructions for debugging and analysis
+use std::fmt;
+
+/// Type of RISC-V instruction
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum InstructionType {
+    // R-type (register-register operations)
+    Add,
+    Sub,
+    Sll,
+    Slt,
+    Sltu,
+    Xor,
+    Srl,
+    Sra,
+    Or,
+    And,
+
+    // I-type (immediate operations)
+    Addi,
+    Slti,
+    Sltiu,
+    Xori,
+    Ori,
+    Andi,
+    Slli,
+    Srli,
+    Srai,
+
+    // Load instructions
+    Lw,
+    Lh,
+    Lb,
+    Lhu,
+    Lbu,
+
+    // Store instructions
+    Sw,
+    Sh,
+    Sb,
+
+    // Branch instructions
+    Beq,
+    Bne,
+    Blt,
+    Bge,
+    Bltu,
+    Bgeu,
+
+    // Upper immediate
+    Lui,
+    Auipc,
+
+    // Jump instructions
+    Jal,
+    Jalr,
+
+    // Unknown instruction
+    Unknown,
+}
+
+/// Operand for an instruction.
+///
+/// Currently, all decoded operands are modeled as [`Operand::Register`] and any
+/// immediate associated with the instruction is stored separately in
+/// [`InstructionTrace::immediate`]. The [`Operand::Immediate`] variant is kept
+/// for potential future use (e.g. representing all operands uniformly in one
+/// place) but is not used by the tracing logic at this time.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Operand {
+    /// Register operand with register number and value
+    Register { reg: u8, value: u32 },
+    /// Immediate value (signed).
+    ///
+    /// Note: The trace format currently uses [`InstructionTrace::immediate`]
+    /// to carry immediate values instead of this variant. This variant is
+    /// reserved for potential future extensions and may not be constructed
+    /// by the existing code.
+    Immediate(i32),
+}
```

Comment:
Let's just rename this to RegisterOperand and convert it to a normal structure instead of an enum. If we're using a separate immediate field anyways then we don't need this complexity.